File: SpEditText/src/main/java/com/sunhapper/x/spedit/gif/span/ResizeIsoheightImageSpan.java
package com.sunhapper.x.spedit.gif.span;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.NonNull;

import com.sunhapper.x.spedit.gif.drawable.ResizeDrawable;


public class ResizeIsoheightImageSpan extends GifIsoheightImageSpan implements RefreshSpan {


    public ResizeIsoheightImageSpan(Drawable d) {
        super(d);
    }

    public ResizeIsoheightImageSpan(@NonNull Drawable d, int verticalAlignment) {
        super(d, verticalAlignment);
    }

    public ResizeIsoheightImageSpan(@NonNull Drawable d, @NonNull String source) {
        super(d, source);
    }

    public ResizeIsoheightImageSpan(@NonNull Drawable d, @NonNull String source, int verticalAlignment) {
        super(d, source, verticalAlignment);
    }

    public ResizeIsoheightImageSpan(Context context, Uri uri) {
        super(context, uri);
    }

    public ResizeIsoheightImageSpan(@NonNull Context context, @NonNull Uri uri, int verticalAlignment) {
        super(context, uri, verticalAlignment);
    }

    public ResizeIsoheightImageSpan(@NonNull Context context, int resourceId) {
        super(context, resourceId);
    }

    public ResizeIsoheightImageSpan(@NonNull Context context, int resourceId, int verticalAlignment) {
        super(context, resourceId, verticalAlignment);
    }

    public ResizeIsoheightImageSpan(@NonNull Bitmap b) {
        super(b);
    }

    public ResizeIsoheightImageSpan(@NonNull Bitmap b, int verticalAlignment) {
        super(b, verticalAlignment);
    }

    public ResizeIsoheightImageSpan(@NonNull Context context, @NonNull Bitmap b) {
        super(context, b);
    }

    public ResizeIsoheightImageSpan(@NonNull Context context, @NonNull Bitmap b, int verticalAlignment) {
        super(context, b, verticalAlignment);
    }

    @Override
    protected Drawable getResizedDrawable() {
        Drawable d = getDrawable();
        if (drawableHeight == 0) {
            return d;
        }
        if (d instanceof ResizeDrawable) {
            ResizeDrawable resizeDrawable = (ResizeDrawable) d;
            if (resizeDrawable.needResize() || !resized) {
                resized = true;
                d.setBounds(new Rect(0, 0,
                        (int) (1f * drawableHeight * resizeDrawable.getWidth() / resizeDrawable.getHeight()),
                        drawableHeight));
            }
        } else if (!resized) {
            resized = true;
            d.setBounds(new Rect(0, 0,
                    (int) (1f * drawableHeight * d.getIntrinsicWidth() / d.getIntrinsicWidth()),
                    drawableHeight));
        }
        return d;
    }

}

File: SpEditText/src/main/java/com/sunhapper/x/spedit/gif/span/ResizeIsoheightImageSpan.kt
package com.sunhapper.x.spedit.gif.span

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.net.Uri

import com.sunhapper.x.spedit.gif.drawable.ResizeDrawable


class ResizeIsoheightImageSpan : GifIsoheightImageSpan, RefreshSpan {


    constructor(d: Drawable) : super(d)

    constructor(d: Drawable, verticalAlignment: Int) : super(d, verticalAlignment)

    constructor(d: Drawable, source: String) : super(d, source)

    constructor(d: Drawable, source: String, verticalAlignment: Int) : super(d, source, verticalAlignment)

    constructor(context: Context, uri: Uri) : super(context, uri)

    constructor(context: Context, uri: Uri, verticalAlignment: Int) : super(context, uri, verticalAlignment)

    constructor(context: Context, resourceId: Int) : super(context, resourceId)

    constructor(context: Context, resourceId: Int, verticalAlignment: Int) : super(context, resourceId, verticalAlignment)

    constructor(b: Bitmap) : super(b)

    constructor(b: Bitmap, verticalAlignment: Int) : super(b, verticalAlignment)

    constructor(context: Context, b: Bitmap) : super(context, b)

    constructor(context: Context, b: Bitmap, verticalAlignment: Int) : super(context, b, verticalAlignment)

    override fun getResizedDrawable(): Drawable {
        val d = drawable
        if (drawableHeight == 0) {
            return d
        }
        if (d is ResizeDrawable && (d.needResize || !resized)) {
            resizeSpan(d)
        } else if (!resized) {
            resizeSpan(d)
        }
        return d
    }

    private fun resizeSpan(d: Drawable) {
        resized = true
        d.bounds = Rect(0, 0,
                (1f * drawableHeight * d.intrinsicWidth / d.intrinsicHeight).toInt(),
                drawableHeight)
    }

}
